### PR TITLE
Fix infinite gas estimate for library bound function call

### LINKED ISSUE #4471: Gas estimates infinite when using libraries
The compilation of the following contract results that the gas estimate of the function `Test.increment` is infinite.

```
pragma solidity 0.4.24;

library Math {
    function add(uint256 a, uint256 b) internal pure returns (uint256 c) {
        return a + b;
    }
}

contract Test {
    using Math for uint;
    
    uint val;
        
    function increment(uint delta) public {
        val = val.add(delta);   
    }
}
```

Solc version
-------------
0.4.24+commit.e67f0147.Emscripten.clang

Reproduce steps
------------------
Use Remix, copy the above contract code there, see a blue warning ("Gas requirement of function Test.increment(uint256) high: infinite").

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #4471

### CODE REVIEW

**Comment by chriseth** [comment]
Can you explain the changes in code generated? If this is just to make the gas estimator happy, I'm not sure we should do that.

**Comment by axic** [comment]
> Can you explain the changes in code generated?

It is a fun reason, we went through this with @leonardoalt.

In short: in the old version `label` is swapped a couple of times and the estimator is not able to deduce it is a label due to the swaps. With reordering the code (I must admit I haven't reviewed the final version of the PR) to avoid swaps (as much as we can), while making the generator code more cohesive (subjective I guess), the estimator is also able to pick it up.

Alternatively the estimator could be improved.

**Comment by leonardoalt** [comment]
The basic idea of this code is to push the jump target right before the jump.
Complementing what @axic said, there are a few `swap`s to organize the stack when the function is `bound`. Since the jump target is currently pushed before `self` (the variable), the Gas Estimator fails to infer the jump target (because of the `swap`s that organize the stack).

**Comment by axic** [comment]
@chriseth @leonardoalt can we try to get this into 0.5.1?

@ekpyron @christianparpart @bit-shift can you also review?

**Comment by leonardoalt** [comment]
I will take another look at this code.

File: Changelog.md [hunk 1]
```diff
@@ -112,6 +112,7 @@ Bugfixes:
  * Type Checker: Report error when using indexed structs in events with experimental ABIEncoderV2. This used to log wrong values.
  * Type Checker: Dynamic types as key for public mappings return error instead of assertion fail.
  * Type System: Allow arbitrary exponents for literals with a mantissa of zero.
+ * Gas Estimator: Fix infinite estimate for bounded library function calls.
```
Comment (ekpyron) [inline Changelog.md:115]: Needs to be moved to 0.5.1 (and the PR rebased I guess).

**Review by ekpyron — CHANGES_REQUESTED** [review]

**Comment by leonardoalt** [comment]
The current code in this PR delays pushing the function label (in `FunctionCall` instead of `MemberAccess`) such that the gas estimator can figure out where to jump to. However, if there is no function call (see file `bound_no_call.sol` in this PR), too many pops are performed at the end, since
-  `a.double` has type `FunctionType` and the `sizeOnStack` of that type is the size of the label + the size of the first parameter.
- the function label was never pushed

**Comment by leonardoalt** [comment]
Closing in favor of doing it properly with Yul.

**Pull request closed without merging**